### PR TITLE
feat: [CDS-71998]: support service variables as env vars in shell script

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
1. Fetch service variables from sweeping output and merge with existing input vars 
2. Render the variables since service vars are not rendered in service step

Below is the output of printenv command in shell script running on remote host.
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49316)

### CODE REVIEW

**Comment by hingerabhinav** [comment]
trigger uts

File: 120-ng-manager/modules/ng-settings/contracts/src/main/java/io/harness/ngsettings/SettingIdentifiers.java [hunk 1]
```diff
@@ -31,4 +31,5 @@ public class SettingIdentifiers {
   public static String INCLUDE_GCP_DISCOUNTS_IDENTIFIER = "include_gcp_discounts";
   public static String INCLUDE_GCP_TAXES_IDENTIFIER = "include_gcp_taxes";
   public static String SERVICE_OVERRIDE_V2_IDENTIFIER = "service_override_v2";
+  public static String EXPORT_SERVICE_VARS_AS_ENV_VARS = "export_service_variables_as_env_variables";
```
Comment (yogesh-chauhan) [inline 120-ng-manager/modules/ng-settings/contracts/src/main/java/io/harness/ngsettings/SettingIdentifiers.java:34]: can the description of the setting indicate that this is only for ssh ?

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java [hunk 2]
```diff
@@ -59,6 +68,7 @@
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.lang3.StringUtils;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java:71]: lets not use Nullable from jetbrains. We can use the lombok one or javax

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java [hunk 3]
```diff
@@ -188,6 +234,15 @@
     return commandPath;
   }
 
+  @VisibleForTesting
+  protected boolean isExportServiceVarsAsEnvVarsEnabled(
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java:238]: can this be private ?

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java [hunk 4]
```diff
@@ -69,30 +79,66 @@
...
       }
     });
+
+    engineExpressionService.resolve(
+        ambiance, res, ExpressionMode.RETURN_ORIGINAL_EXPRESSION_IF_UNRESOLVED, new HashMap<>());
+
+    // no harness expressions should remain
+    res.forEach((key, value) -> {
+      if (EngineExpressionEvaluator.hasExpressions(value)) {
+        throw new InvalidRequestException(String.format("Env. variable [%s] value found to be unresolved", key));
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java:120]: just to confirm, if there is an unresolved expression in input variables, we fail today right ? Looks like the case checking at this line

```java
if (parameterFieldValue.getValue() == null) {
          throw new InvalidRequestException(String.format("Env. variable [%s] value found to be null", key));
        }
```
But just to make sure

**Comment by hingerabhinav** [comment]
trigger uts

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49316)* status: 4 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-settings/contracts/src/main/java/io/harness/ngsettings/SettingIdentifiers.java` line 34 at r2](https://reviewable.io/reviews/harness/harness-core/49316#gh-1233749352:-NYMfOfO7VmWRdrIPivy:bu5j373) ([raw file](https://github.com/harness/harness-core/blob/b9143c07499796f9caedca8f32bbd552e89c0b64/120-ng-manager/modules/ng-settings/contracts/src/main/java/io/harness/ngsettings/SettingIdentifiers.java#L34)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can the description of the setting indicate that this is only for ssh ?
</blockquote></details>

This feature works for both bash (delegate and remote) and powershell (delegate and remote). 
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java` line 71 at r2](https://reviewable.io/reviews/harness/harness-core/49316#gh-1233752449:-NYMfZPP5SvATqaQ2Szs:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b9143c07499796f9caedca8f32bbd552e89c0b64/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java#L71)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets not use Nullable from jetbrains. We can use the lombok one or javax
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java` line 119 at r2](https://reviewable.io/reviews/harness/harness-core/49316#gh-1233762060:-NYMf_07-nF8MI1-b39k:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b9143c07499796f9caedca8f32bbd552e89c0b64/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java#L119)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should we not check all values which are unresolved and throw exception with all those variable names ?
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java` line 120 at r2](https://reviewable.io/reviews/harness/harness-core/49316#gh-1233760776:-NYMfaXe-91j6Sr7ipwl:byvg4h8) ([raw file](https://github.com/harness/harness-core/blob/b9143c07499796f9caedca8f32bbd552e89c0b64/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java#L120)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

just to confirm, if there is an unresolved expression in input variables, we fail today right ? Looks like the case checking at this line

```java
if (parameterFieldValue.getValue() == null) {
          throw new InvalidRequestException(String.format("Env. variable [%s] value found to be null", key));
        }
```
But just to make sure
</blockquote></details>

Yes. Added a check after this processing. Done. 
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java` line 238 at r2](https://reviewable.io/reviews/harness/harness-core/49316#gh-1233753028:-NYMfeI8CtS-Zoj6fl4O:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b9143c07499796f9caedca8f32bbd552e89c0b64/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/shellscript/ShellScriptHelperServiceImpl.java#L238)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can this be private ?
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49316)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by nishantsaini20 — APPROVED** [review]
Approved for SettingIdentifiers

**Review by puthrayaharness — APPROVED** [review]

**Comment by hingerabhinav** [comment]
trigger review

**Pull request merged**